### PR TITLE
unix: signal done to last thread barrier waiter

### PR DESCRIPTION
**Pull request by bnoordhuis**
Libuv's own thread barrier implementation signaled completion to the
first waiter that saw the threshold being reached, contrary to what
some native pthreads barrier implementations do, which is to signal
it to the _last_ waiter.

Libuv's behavior is not strictly non-conforming but it's inconvenient
because it means this snippet (that appears in the libuv documentation)
has a race condition in it:

    if (uv_barrier_wait(&barrier) > 0)
      uv_barrier_destroy(&barrier);  // can still have waiters

This issue was discovered and fixed by Ali Ijaz Sheikh, a.k.a @ofrobots,
but some refactoring introduced conflicts in his pull request and I
didn't have the heart to ask him to redo it from scratch. :-)

cc @davisjam @ofrobots, and cc @jBarz for the AIX commit.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1049/

### CODE REVIEW

File: src/unix/thread.c [hunk 1]
```diff
@@ -44,108 +44,119 @@
...
-int pthread_barrier_destroy(pthread_barrier_t* barrier) {
-  int rc;
-  _uv_barrier* b;
 
-  if (barrier == NULL || barrier->b == NULL)
-    return EINVAL;
+void uv_barrier_destroy(uv_barrier_t* barrier) {
+  struct _uv_barrier* b;
 
   b = barrier->b;
```
Comment (santigimeno) [inline src/unix/thread.c]: Maybe add an assert here that `barrier` and `barrier->b` are not `NULL`?

File: test/test-barrier.c [hunk 2]
```diff
@@ -104,3 +104,46 @@ TEST_IMPL(barrier_3) {
 
   return 0;
 }
+
+static void serial_worker(void* data) {
+  uv_barrier_t* barrier = data;
+  if (uv_barrier_wait(barrier) > 0) {
+    uv_barrier_destroy(barrier);
+  }
```
Comment (santigimeno) [inline test/test-barrier.c]: style: remove braces

File: test/test-barrier.c [hunk 3]
```diff
@@ -104,3 +104,46 @@ TEST_IMPL(barrier_3) {
...
+    ASSERT(0 == uv_thread_create(&threads[i], serial_worker, &barrier));
+  }
+
+  if (uv_barrier_wait(&barrier) > 0) {
+    uv_barrier_destroy(&barrier);
+  }
+
+  for (i = 0; i < ARRAY_SIZE(threads); ++i) {
+    ASSERT(0 == uv_thread_join(&threads[i]));
+  }
```
Comment (santigimeno) [inline test/test-barrier.c]: Remove braces here too

File: test/test-barrier.c [hunk 4]
```diff
@@ -104,3 +104,46 @@ TEST_IMPL(barrier_3) {
 
   return 0;
 }
+
+static void serial_worker(void* data) {
+  uv_barrier_t* barrier = data;
```
Comment (santigimeno) [inline test/test-barrier.c]: Style: separate declaration from assignment?

File: src/unix/thread.c [hunk 5]
```diff
@@ -44,108 +44,119 @@
...
+  uv_mutex_lock(&b->mutex);
 
-  if ((rc = pthread_mutex_lock(&b->mutex)) != 0)
-    return rc;
+  assert(b->in == 0);
```
Comment (davisjam) [inline src/unix/thread.c]: I think it's worth adding a comment here saying that asserts like these interpret "undefined behavior" in the spec as a crash.

File: src/unix/thread.c [hunk 6]
```diff
@@ -44,108 +44,119 @@
...
+  assert(b->out == 0);
 
-  if (b->in > 0 || b->out > 0)
-    rc = EBUSY;
-
-  pthread_mutex_unlock(&b->mutex);
+  if (b->in != 0 || b->out != 0)
```
Comment (davisjam) [inline src/unix/thread.c]: This condition and the asserts immediately preceding accomplish the same thing.

Comment (santigimeno) [reply src/unix/thread.c]: `assert()` does not generate code when run in release mode (with `NDEBUG` set)

Comment (davisjam) [reply src/unix/thread.c]: If we're going to abort with or without `NDEBUG`, why bother asserting?

Comment (santigimeno) [reply src/unix/thread.c]: I would say better error message

File: src/unix/thread.c (hunk 1 above)
Comment (santigimeno) [reply src/unix/thread.c]: No strong feelings either. Your choice.

File: src/unix/thread.c (hunk 5 above)
Comment (bnoordhuis) [reply src/unix/thread.c]: Santiago already commented elsewhere but to reiterate, we use asserts as sanity checks in debug builds only. It's still truly UB in release builds.

**Comment by bnoordhuis** [comment]
Landed in daf04e83...3af6f172. I left out the asserts.

**Pull request closed without merging**